<a target="_blank" href="https://colab.research.google.com/github/lukebarousse/Python_Data_Analytics_Course/blob/main/2_Advanced/02_Pandas_Data_Cleaning.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

# Pandas Data Cleaning

Load data.

In [9]:
# Importing Libraries
import pandas as pd
from datasets import load_dataset
import matplotlib.pyplot as plt  

# Loading Data
dataset = load_dataset('lukebarousse/data_jobs')
df = dataset['train'].to_pandas()

# Data Cleanup
df['job_posted_date'] = pd.to_datetime(df['job_posted_date'])

## Handling Missing Data

### Review

This is what we learned in the basics section, this is just a refresher of how we've handled null values before. Feel free to skip this.

#### Notes

- `df.dropna()`: Drop missing values.

##### Examples

Here we are only drop values if all of their values are missing. 

In [10]:
df_cleaned = df.dropna(how='all')
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 787686 entries, 0 to 787685
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   job_title_short        787686 non-null  object 
 1   job_title              787685 non-null  object 
 2   job_location           786638 non-null  object 
 3   job_via                787678 non-null  object 
 4   job_schedule_type      774976 non-null  object 
 5   job_work_from_home     787686 non-null  bool   
 6   search_location        787686 non-null  object 
 7   job_posted_date        787686 non-null  object 
 8   job_no_degree_mention  787686 non-null  bool   
 9   job_health_insurance   787686 non-null  bool   
 10  job_country            787636 non-null  object 
 11  salary_rate            33089 non-null   object 
 12  salary_year_avg        22046 non-null   float64
 13  salary_hour_avg        10648 non-null   float64
 14  company_name           787669 non-nu

Right now all we can do is drop values if they're missing. But that's not useful right now because our DataFrame didn't have any. 

So, what if we wanted to fill the missing values with something else? This is expecially useful so we don't run into errors when dealing with NaN values.

### Fillna

#### Notes

- `df.fillna()`: Fill missing values

#### Examples

Let's fill in instances where there's no salary info (aka these columns have NaN values `salary_rate`, `salary_year_avg`, `salary_hour_avg`) with 0.

We're going to look at a few rows in these 3 columns right now, so we can compare what we've done before to after. 

We'll use `iloc` to look at the first 10 rows `:20` and the salary information rows `11:14`.

In [11]:
df_cleaned.iloc[:10,11:14]

salary_rate  salary_year_avg  salary_hour_avg
0         NaN              NaN              NaN
1         NaN              NaN              NaN
2        year         425000.0              NaN
3         NaN              NaN              NaN
4         NaN              NaN              NaN
5         NaN              NaN              NaN
6        year         180000.0              NaN
7         NaN              NaN              NaN
8         NaN              NaN              NaN
9         NaN              NaN              NaN

We fill the values for the 3 columns with 0 using `fillna()`.

In [12]:
fill_values = ['salary_rate', 'salary_year_avg', 'salary_hour_avg']
df_filled = df_cleaned.fillna(0)
df_filled.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 787686 entries, 0 to 787685
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   job_title_short        787686 non-null  object 
 1   job_title              787686 non-null  object 
 2   job_location           787686 non-null  object 
 3   job_via                787686 non-null  object 
 4   job_schedule_type      787686 non-null  object 
 5   job_work_from_home     787686 non-null  bool   
 6   search_location        787686 non-null  object 
 7   job_posted_date        787686 non-null  object 
 8   job_no_degree_mention  787686 non-null  bool   
 9   job_health_insurance   787686 non-null  bool   
 10  job_country            787686 non-null  object 
 11  salary_rate            787686 non-null  object 
 12  salary_year_avg        787686 non-null  float64
 13  salary_hour_avg        787686 non-null  float64
 14  company_name           787686 non-nu

Now if we compare the results using `iloc` again on our new DataFrame. We see that the previous NaN values in the columns ( `salary_rate`, `salary_year_avg`, `salary_hour_avg`) have been replaced with 0.

In [13]:
df_filled.iloc[:10,11:14]

salary_rate  salary_year_avg  salary_hour_avg
0           0              0.0              0.0
1           0              0.0              0.0
2        year         425000.0              0.0
3           0              0.0              0.0
4           0              0.0              0.0
5           0              0.0              0.0
6        year         180000.0              0.0
7           0              0.0              0.0
8           0              0.0              0.0
9           0              0.0              0.0

## Drop Duplicates

### Notes

* `drop_duplicates()`: Remove duplicate rows.
* Analysts will often need to clean up data and one of the most common issues we run into is duplicate values. 

### Examples

Now that we've dealt with NaN values. Let's continue cleaning the data by removing any duplicate rows. 

In [14]:
df_unique = df_filled.drop_duplicates()
df_unique.info()

<class 'pandas.core.frame.DataFrame'>
Index: 787573 entries, 0 to 787685
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   job_title_short        787573 non-null  object 
 1   job_title              787573 non-null  object 
 2   job_location           787573 non-null  object 
 3   job_via                787573 non-null  object 
 4   job_schedule_type      787573 non-null  object 
 5   job_work_from_home     787573 non-null  bool   
 6   search_location        787573 non-null  object 
 7   job_posted_date        787573 non-null  object 
 8   job_no_degree_mention  787573 non-null  bool   
 9   job_health_insurance   787573 non-null  bool   
 10  job_country            787573 non-null  object 
 11  salary_rate            787573 non-null  object 
 12  salary_year_avg        787573 non-null  float64
 13  salary_hour_avg        787573 non-null  float64
 14  company_name           787573 non-null  o

If you compare that with the original DataFrame which had 787686 entries. This new DataFrame `df_unique` has 787577 entries. It removed 109 entries. 

Now let's see what would happen if we tried to remove duplicates from `job_title`.

In [15]:
df_unique = df_filled.drop_duplicates(subset=['job_title'])
df_unique.info()

<class 'pandas.core.frame.DataFrame'>
Index: 235063 entries, 0 to 787685
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   job_title_short        235063 non-null  object 
 1   job_title              235063 non-null  object 
 2   job_location           235063 non-null  object 
 3   job_via                235063 non-null  object 
 4   job_schedule_type      235063 non-null  object 
 5   job_work_from_home     235063 non-null  bool   
 6   search_location        235063 non-null  object 
 7   job_posted_date        235063 non-null  object 
 8   job_no_degree_mention  235063 non-null  bool   
 9   job_health_insurance   235063 non-null  bool   
 10  job_country            235063 non-null  object 
 11  salary_rate            235063 non-null  object 
 12  salary_year_avg        235063 non-null  float64
 13  salary_hour_avg        235063 non-null  float64
 14  company_name           235063 non-null  o

If we look at it now. It looks like we removed quite a few rows. Now all of these rows have unique `job_title`s. 

In [16]:
df_unique.head(10)

job_title_short                                          job_title  \
0         Data Engineer      DevOps Engineer - Big Data/Advanced Analytics   
1         Data Engineer     Co-Op/Intern Software Engineer, Data Ingestion   
2         Data Engineer                             Hybrid - Data Engineer   
3        Data Scientist                                     Data Scientist   
4         Data Engineer  Data Lead Engineer (with strong Python) - Remo...   
5  Senior Data Engineer                               Senior Data Engineer   
6         Data Engineer                                      Data Engineer   
7          Data Analyst      Applied Mathematician, Scientist, or Engineer   
8  Senior Data Engineer  Sr. Big Data Engineer with Data Bricks, Loc - ...   
9         Data Engineer                                Azure Data Engineer   

                             job_location            job_via  \
0  Bari, Metropolitan City of Bari, Italy       via LinkedIn   
1                   Vancouver, BC, Canada       via LinkedIn   
2                            New York, NY       via LinkedIn   
3                           Hamtramck, MI          via BeBee   
4                                Anywhere      via Jobgether   
5                             Milford, CT       via Snagajob   
6                                   Texas       via LinkedIn   
7                     Copenhagen, Denmark  via BeBee Danmark   
8                              Quincy, MA   via ZipRecruiter   
9                              Warren, NJ       via LinkedIn   

  job_schedule_type  job_work_from_home            search_location  \
0         Full-time               False                      Italy   
1         Full-time               False                     Canada   
2         Full-time               False                      Sudan   
3         Full-time               False    Illinois, United States   
4         Full-time                True                     Panama   
5         Full-time               False  California, United States   
6         Full-time               False       Texas, United States   
7         Full-time               False                    Denmark   
8         Full-time               False  California, United States   
9        Contractor               False  California, United States   

       job_posted_date  job_no_degree_mention  job_health_insurance  \
0  2023-10-24 18:16:15                   True                 False   
1  2023-05-05 18:37:58                  False                 False   
2  2023-07-25 19:07:08                  False                 False   
3  2023-11-30 18:05:20                  False                 False   
4  2023-08-13 18:32:30                  False                 False   
5  2023-07-28 18:06:44                  False                 False   
6  2023-03-06 18:51:56                  False                  True   
7  2023-07-12 18:24:10                  False                 False   
8  2023-05-25 18:05:51                   True                 False   
9  2023-09-20 18:05:42                   True                 False   

     job_country salary_rate  salary_year_avg  salary_hour_avg  \
0          Italy           0              0.0              0.0   
1         Canada           0              0.0              0.0   
2          Sudan        year         425000.0              0.0   
3  United States           0              0.0              0.0   
4         Panama           0              0.0              0.0   
5  United States           0              0.0              0.0   
6  United States        year         180000.0              0.0   
7        Denmark           0              0.0              0.0   
8  United States           0              0.0              0.0   
9  United States           0              0.0              0.0   

                        company_name  \
0                    NTT DATA Italia   
1                            Kinaxis   
2                  Durlston Partners   
3   

For our example we don't really need to remove any duplicates right now, but it's important to understand the concept.